### PR TITLE
Migrate main MeshPy functionality to YAML

### PR DESCRIPTION
**Pull request by isteinbrecher**
This PR is a first shot at getting MeshPy ready for YAML.

For now the previous interfaces stay mostly the same, only some things, e.g., boundary conditions, have to be entered differently (see the changes in the test files).

At the moment we use the default `yaml` library to handle read and dump operations, but that could/should change in the future, once `fourcipp` is ready.

Currently almost all tests pass, except the cubitpy tests, the header function tests and the beam potential tests.

The main (and actually only real) functional change is in the `InputFile` where the whole "convert plain input data to MeshPy" and "convert MeshPy to plain input data" was refactored and cleaned up.

This turned out to be quite a large PR - however many changes are just repetitive testing changes. Let's see what the best way forward is, I would also be fine in merging this file after file to simplify the review process.

### CODE REVIEW

File: src/meshpy/core/boundary_condition.py [hunk 1]
```diff
@@ -74,22 +71,25 @@ def from_dat(cls, bc_key, line, **kwargs):
                 BoundaryCondition as _BoundaryCondition,
             )
 
-            return _BoundaryCondition(
-                int(split[1]) - 1, " ".join(split[2:]), bc_type=bc_key, **kwargs
+            boundary_condition = _BoundaryCondition(
+                geometry_set, bc_dict, bc_type=bc_key
```
Comment (davidrudlstorfer) [inline src/meshpy/core/boundary_condition.py:75]: I must say that this is so much nicer and more straightforward than before

File: src/meshpy/core/element_beam.py [hunk 2]
```diff
@@ -155,18 +155,18 @@ def check_node(node, pos, rot, arc_length, name):
         return created_nodes
 
     @classmethod
-    def get_coupling_string(cls, coupling_dof_type):
-        """Return the string to couple this beam to another beam."""
+    def get_coupling_dict(cls, coupling_dof_type):
```
Comment (davidrudlstorfer) [inline src/meshpy/core/element_beam.py:158]: should we rename all of these functions to something more generic?

i.e., `get_coupling_input` or `get_coupling_data`


This could also be done in a future PR

File: src/meshpy/core/element_volume.py [hunk 3]
```diff
@@ -36,23 +36,21 @@ class VolumeElement(_Element):
...
-            nodes_string += f"{node.i_global} "
+        nodes_string = " ".join(str(node.i_global) for node in self.nodes)
 
-        # Return the dat line.
-        return (
-            f"{self.i_global} {self.dat_pre_nodes} {nodes_string} {self.dat_post_nodes}"
-        )
+        return [
+            f"{self.i_global} {self.string_pre_nodes} {nodes_string} {self.string_post_nodes}"
+        ]
```
Comment (davidrudlstorfer) [inline src/meshpy/core/element_volume.py:53]: Why do we actually dump this into a list here?

If I understand it correctly with fourcipp we can then change this back to a dict?

File: src/meshpy/cosserat_curve/warping_along_cosserat_curve.py [hunk 4]
```diff
@@ -296,12 +296,17 @@ def create_transform_boundary_conditions(
...
                 _GeometrySet(node),
-                "NUMDOF {4} ONOFF 1 1 1 {3}VAL 1.0 1.0 1.0 {3}FUNCT {0} {1} {2} {3}TAG monitor_reaction",
-                format_replacement=fun_pos + [additional_dof, n_dof_per_node],
+                {
+                    "NUMDOF": n_dof_per_node,
+                    "ONOFF": [1] * 3 + [0] * n_additional_dof,
+                    "VAL": [1.0] * 3 + [0.0] * n_additional_dof,
+                    "FUNCT": fun_pos + [None] * n_additional_dof,
+                    "TAG": "monitor_reaction",
+                },
```
Comment (davidrudlstorfer) [inline src/meshpy/cosserat_curve/warping_along_cosserat_curve.py:309]: really nice changes:D

File: src/meshpy/four_c/beam_potential.py
Comment (davidrudlstorfer) [inline src/meshpy/four_c/beam_potential.py:1]: Thanks for also directly tackling this!

File: src/meshpy/four_c/input_file.py [hunk 5]
```diff
@@ -22,224 +22,83 @@
...
-                return
-            data_lines = data
-
-        # Remove the first and or last line if it is empty.
-        for index in [0, -1]:
-            if data_lines[index].strip() == "":
-                del data_lines[index]
-
-        # Add the data lines.
-        for item in data_lines:
```
Comment (davidrudlstorfer) [inline src/meshpy/four_c/input_file.py:199]: This removal is really satisfying :D

File: src/meshpy/four_c/yaml_dumper.py
Comment (davidrudlstorfer) [inline src/meshpy/four_c/yaml_dumper.py:1]: With fourcipp this can go away too?

File: src/meshpy/utils/environment.py
Comment (davidrudlstorfer) [inline src/meshpy/utils/environment.py:1]: This is currently false for the pipeline because we have not yet installed the optional dependency, correct?

File: tests/reference-files/test_four_c_simulation_dirichlet_boundary_to_neumann_boundary_with_all_values_neumann.4C.yaml [hunk 6]
```diff
@@ -25,21 +32,21 @@ RESULT DESCRIPTION:
...
+      TOLERANCE: 1.e-10
   - STRUCTURE:
       DIS: "structure"
       NODE: 21
       QUANTITY: "dispz"
       VALUE: 0
-      TOLERANCE: 1e-10
+      TOLERANCE: 1.e-10
 STRUCT NOX/Status Test:
-  XML File: "/home/rudlstorfer/01_work/01_workspace/meshpy/tests/reference-files/NOT_DEFINED"
```
Comment (davidrudlstorfer) [inline tests/reference-files/test_four_c_simulation_dirichlet_boundary_to_neumann_boundary_with_all_values_neumann.4C.yaml:42]: Thanks for removing these

**Review by davidrudlstorfer — COMMENTED** [review]
@isteinbrecher thanks a lot for all of the work you already put into this to move MeshPy to yaml based input files for 4C

I've looked through all code changes and in my opinion everything is looking very nice already.

I just left a few questions and ideas - but nothing big.

With fourcipp we will be able to further simplify a lot of stuff. I am really looking forward to it

Let's discuss my open comments in our quick meeting today and then we can proceed here.

**Comment by isteinbrecher** [comment]
With the latest pushes, all tests except the beam-potential, header tests, and all cubitpy related tests should pass.

I will have another close look again at the changes, and I would especially ask @davidrudlstorfer and @knarfnitram to also do so, then I think we can merge this (to `mograte_dat_to_yaml`) within the next days.

File: src/meshpy/four_c/beam_interaction_conditions.py [hunk 7]
```diff
@@ -71,15 +67,12 @@ def get_next_possible_id_for_boundary_condition(
...
-                raise ValueError(
-                    "The string provided "
-                    + regex_search_string
-                    + " could not be found within the condition "
-                    + bc.bc_string
+            if condition_string in bc.bc_dict:
+                existing_ids.append(bc.bc_dict[condition_string])
+            else:
+                raise KeyError(
+                    f"The key {condition_string} is not in the bc_dict {bc.bc_dict}"
```
Comment (knarfnitram) [inline src/meshpy/four_c/beam_interaction_conditions.py:74]: that's a nice simplification!

File: src/meshpy/four_c/header_functions.py [hunk 8]
```diff
@@ -111,83 +110,90 @@ def set_runtime_output(
...
-            option_overwrite=option_overwrite,
-        )
+        {
+            "IO/RUNTIME VTK OUTPUT/STRUCTURE": {
+                "OUTPUT_STRUCTURE": output_solid,
+                "DISPLACEMENT": True,
+                "STRESS_STRAIN": output_stress_strain,
+                "ELEMENT_OWNER": element_owner,
+                "ELEMENT_GID": element_gid,
+                "ELEMENT_MAT_ID": element_mat_id,
```
Comment (knarfnitram) [inline src/meshpy/four_c/header_functions.py:132]: Since you removed `get_yes_no()`, I am wondering if we could remove this function completely?

File: src/meshpy/four_c/header_functions.py [hunk 9]
```diff
@@ -736,3 +709,37 @@ def set_beam_contact_section(
...
+
+
+def add_result_description(
+    input_file: _InputFile,
+    displacements: _List,
+    node_ids: _List[int],
+    *,
+    tol: float = 1e-10,
+):
+    """Add result descriptions to the input file.
```
Comment (knarfnitram) [inline src/meshpy/four_c/header_functions.py:721]: """Adds result descriptions for structure problems to the input file

**Comment by knarfnitram** [comment]
Thanks for taking the initiative and switching to `yaml` @isteinbrecher . Some of the previous code passages are now much more simpler!
I  just left some minor comments/ questions. But I think we can merge this rather soon.

File: src/meshpy/four_c/header_functions.py (hunk 8 above)
Comment (isteinbrecher) [reply src/meshpy/four_c/header_functions.py:132]: Good point, I will remove it.

**Comment by isteinbrecher** [comment]
Thanks for all the comments! I had another look myself and this turned out to be surprisingly mature 😅. @knarfnitram if you are fine with my comments and changes, I think we can merge this.

**Review by davidrudlstorfer — APPROVED** [review]

**Review by knarfnitram — APPROVED** [review]

**Comment by davidrudlstorfer** [comment]
Let's get this merged :rocket: thanks again

Once merged I'll fix the beam potential stuff and add the new header to the title section as discussed

**Comment by isteinbrecher** [comment]
Thanks for all the input, I really like the improvements that came with this switch, lets see how to further improve on this.
